### ISSUE TITLE
No gradient for `residuals` in the return value of `torch.linalg.lstsq`

### ISSUE THREAD

**Issue by Bichidian** [pytorch/pytorch#147543]
The return value of `torch.linalg.lstsq` is a named tuple `(solution, residuals, rank, singular_values)`. I find that `solution` has gradient but `residuals` does not. Is this expected? I'm using `gels` driver.

cc @ezyang @albanD @gqchen @pearu @nikitaved @soulitzer @Varal7 @xmfan @jianyuh @mruberry @walterddr @xwang233 @Lezcano

**Comment by lezcano** [comment]
In my opinion, we shouldn't even be returning the residuals (we just return them in some cases and in others we don't...) so I'm not surprised we don't implement gradients for them...
That being said, you can compute them yourself in the obvious way and the gradients should be correct then.

**Comment by Bichidian** [comment]
@lezcano Computing `residuals` manually from `solution` should give the correct gradient but is highly inefficient, because the gradient for `solution` $$X^\*$$ involves computing and backpropagating through the pseudoinverse. For the `residuals` $$L^*$$ we can make use of the Danskin's theorem, so we have:
```math
\begin{align}
& L(X, A, B) := \Vert AX-B \Vert^2 \\
& X^*(A, B) := \arg \min_X L(X, A, B) \\
& L^*(A, B) := L \left( X^*(A, B), A, B \right) \\
& \frac{\partial{L^*}}{\partial{A}} = \left. \frac{\partial{L}}{\partial{A}} \right|_{X=X^*} = 2 (AX-B) X^T \\
& \frac{\partial{L^*}}{\partial{B}} = \left. \frac{\partial{L}}{\partial{B}} \right|_{X=X^*} = -2 (AX-B)
\end{align}
```
As you can see, the residuals $$L^*$$ has much simpler derivatives. Here is a code that I'm using now:
```
class LstsqResidual(torch.autograd.Function):
    @staticmethod
    def forward(ctx, A, b):
        """
        A: (batch, num_equations, num_statistics)
        b: (batch, num_equations, 1)
        Returns: scalar
        """
        x = torch.linalg.lstsq(A, b).solution
        r = torch.bmm(A, x) - b
        ctx.save_for_backward(x, r)
        return torch.square(r).mean()

    @staticmethod
    def backward(ctx, grad_output):
        x, r = ctx.saved_tensors
        numel = r.numel()
        grad_A = grad_output * 2 * torch.bmm(r, x.transpose(-1, -2)) / numel
        grad_b = grad_output * (-2 * r) / numel
        return grad_A, grad_b
```
Will you be interested in a pull request that adds this feature? If so, should I do it by modifying `tools/autograd/derivatives.yaml`? Thank you!

**Comment by lezcano** [comment]
Fair enough, send a PR with tests and tag me as a reviewer.

**Comment by Bichidian** [comment]
@lezcano Can you give me some advice about how I should proceed? I see that in `tools/autograd/derivatives.yaml` there is this block:
```
- name: linalg_lstsq(Tensor self, Tensor b, float? rcond=None, *, str? driver=None) -> (Tensor solution, Tensor residuals, Tensor rank, Tensor singular_values)
  self, b: linalg_lstsq_backward(grad, self, b, grad_input_mask)
  solution: linalg_lstsq_jvp(self_p, b_p, self_t, b_t)
  output_differentiability: [True, False, False, False]
```
I can modify it to be:
```
- name: linalg_lstsq(Tensor self, Tensor b, float? rcond=None, *, str? driver=None) -> (Tensor solution, Tensor residuals, Tensor rank, Tensor singular_values)
  self, b: linalg_lstsq_backward(grads[0], grads[1], self, b, grad_input_mask)
  solution: linalg_lstsq_jvp(self_p, b_p, self_t, b_t)
  residuals: linalg_lstsq_residuals_jvp(self_p, b_p, self_t, b_t)
  output_differentiability: [True, True, False, False]
```
Then modify `linalg_lstsq_backward` and implement `linalg_lstsq_residuals_jvp` in `pytorch/torch/csrc/autograd/FunctionsManual.cpp`.

But as you pointed out, the `residuals` is not always returned non-empty. As I understand, it is returned only because the underlying LAPACK routine sometimes returns the residuals. Otherwise it is against the purpose to compute it manually using matrix multiplication (although I see discussions in #56483 which advocates for a `compute_residuals` flag). However, the gradient for the `residuals` $$L^\*$$ depends on the term $$A X^\* − B$$ (see above), requiring matrix multiplication. So using `residuals` returned by `torch.linalg.lstsq` when one needs its gradient is not more efficient than something like the above `class LstsqResidual(torch.autograd.Function)`, which only uses the `solution`. What do you think is a good way to implement this gradient? Thank you!

**Comment by lezcano** [comment]
As discussed above, I really think that `lstsq` should only return the solution. Everything else it returns was just a result as a bad API design. I think it's better to let sleeping dogs lay and leave it as is. Then, if anyone wants to use it, they can just implement it themselves as you discussed above, understanding the rather annoying limitations of the current API...

**Comment by Bichidian** [comment]
I understand. Maybe the API can be improved later by adding a `compute_residuals` flag, or having a separate function. But I'm already halfway through it, and I still think it is a useful feature, so I'm going to open a PR.
